src/backend/renderer/multigpu/mod.rs

```diff
@@ -0,0 +1,1755 @@
+    render: RenderDevice<'a, R>,
+    target: Option<&'b mut T::Device>,
+    other_renderers: Vec<&'a mut R::Device>,
+    proxy_framebuffer: std::marker::PhantomData<Target>,
+    log: ::slog::Logger,
+}
+
+impl<'a, 'b, R: GraphicsApi, T: GraphicsApi, Target> AsRef<<R::Device as ApiDevice>::Renderer>
+    for MultiRenderer<'a, 'b, R, T, Target>
+{
+    fn as_ref(&self) -> &<R::Device as ApiDevice>::Renderer {
+        self.render.renderer()
+    }
+}
+
+impl<'a, 'b, R: GraphicsApi, T: GraphicsApi, Target> AsMut<<R::Device as ApiDevice>::Renderer>
+    for MultiRenderer<'a, 'b, R, T, Target>
+{
+    fn as_mut(&mut self) -> &mut <R::Device as ApiDevice>::Renderer {
+        self.render.renderer_mut()
+    }
+}
+
+// Hack for implementing Renderer::render..
+#[derive(Debug)]
+enum RenderDevice<'a, A: GraphicsApi> {
+    Device(&'a mut A::Device),
+    // Hack to avoid lifetime problems in Renderer::render
+    Renderer(*mut <A::Device as ApiDevice>::Renderer, DrmNode),
+}
+
+impl<'a, A: GraphicsApi> RenderDevice<'a, A> {
+    /*
+    fn unwrap_device(&mut self) -> &mut A::Device {
+        match self {
+            RenderDevice::Device(dev) => *dev,
+            RenderDevice::Renderer(_, _) => panic!("unwrap called on RenderDevice::Renderer"),
+        }
+    }
+    */
+
+    fn node(&self) -> &DrmNode {
+        match self {
+            RenderDevice::Device(dev) => dev.node(),
+            RenderDevice::Renderer(_, node) => node,
+        }
+    }
+
+    fn renderer(&self) -> &<A::Device as ApiDevice>::Renderer {
+        match self {
+            RenderDevice::Device(dev) => dev.renderer(),
+            RenderDevice::Renderer(renderer, _) => unsafe { &**renderer },
+        }
+    }
+
+    fn renderer_mut(&mut self) -> &mut <A::Device as ApiDevice>::Renderer {
+        match self {
+            RenderDevice::Device(dev) => dev.renderer_mut(),
+            RenderDevice::Renderer(renderer, _) => unsafe { &mut **renderer },
+        }
+    }
+}
+
+/// [`Frame`](super::Frame) implementation of a [`MultiRenderer`].
+#[derive(Debug)]
+pub struct MultiFrame<R: GraphicsApi, T: GraphicsApi> {
+    node: DrmNode,
+    frame: *mut <<R::Device as ApiDevice>::Renderer as Renderer>::Frame,
+    damage: Vec<Rectangle<i32, Physical>>,
+    // We need this for the associated Error type of the Frame implementation
+    _target: std::marker::PhantomData<T>,
+    log: ::slog::Logger,
+}
+
+impl<R: GraphicsApi, T: GraphicsApi> AsRef<<<R::Device as ApiDevice>::Renderer as Renderer>::Frame>
+    for MultiFrame<R, T>
+{
+    fn as_ref(&self) -> &<<R::Device as ApiDevice>::Renderer as Renderer>::Frame {
+        unsafe { &*self.frame }
+    }
+}
+
+impl<R: GraphicsApi, T: GraphicsApi> AsMut<<<R::Device as ApiDevice>::Renderer as Renderer>::Frame>
+    for MultiFrame<R, T>
+{
+    fn as_mut(&mut self) -> &mut <<R::Device as ApiDevice>::Renderer as Renderer>::Frame {
+        unsafe { &mut *self.frame }
+    }
+}
```

Comment:
Let's justify how these unsafe blocks are not undefined behavior.

Reply:
Added comment explaining